larpandoracontent/LArHelpers/LArMCParticleHelper.cc

```diff
@@ -34,13 +34,44 @@ LArMCParticleHelper::PrimaryParameters::PrimaryParameters() :
     m_minPrimaryGoodViews(2),
     m_selectInputHits(true),
     m_maxPhotonPropagation(2.5f),
-    m_minHitSharingFraction(0.9f)
+    m_minHitSharingFraction(0.9f), 
+    m_foldBackHierarchy(true)
 {
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+bool LArMCParticleHelper::DoesPrimaryMeetCriteria(const MCParticle *const pMCParticle, std::function<bool(const MCParticle *const)> fCriteria)
+{
+    try
```

Comment:
When testing I found that an error occurs when trying to find the primary MCParticle - from memory I think it was a STATUS_CODE_NOT_FOUND but I could be wrong. When using the helper to create a MCParticle -> hits map you have to provide some criteria that you want the particles to meet, these criteria are things like: 'is it a cosmic ray AND is it a primary'. Obviously when not folding back the hierarchy you want to avoid the latter criteria so for each particle I obtain the primary and apply the criteria to the primary. This means that the particle has to have a primary, so if it doesn't then i just say it doesn't meet the criteria. So there is a hidden criteria that the MCParticle has to have a primary, which I think is something I want to have.

Reply:
Turns out that LArMCParticleHelper::GetPrimaryMCParticle can raise an exception if there's a failure to find said primary, or if the input MC info is in an unexpected form, so that MCParticles can have more than one parent. We're at the mercy of the input MC information, so there are some sanity checks, and it's via testing that these will be picked up.